refactor: subscriptions

hrwX opened the pull request:
- Reduced complexity associated with invoice generation function, got rid of `is_postpaid_to_invoice` and `is_prepaid_to_invoice`.
```
def can_generate_new_invoice(self) -> bool:
    if self.cancelation_date:
	    return False
    elif self.generate_invoice_at_period_start and (
	    getdate(frappe.flags.current_date) == getdate(self.current_invoice_start)
	    or self.is_new_subscription()
    ):
	    return True
    elif getdate(frappe.flags.current_date) == getdate(self.current_invoice_end):
	    if self.has_outstanding_invoice() and not self.generate_new_invoices_past_due_date:
		    return False
    
	    return True
    else:
	    return False
```
- The invoices won't be generated randomly, they will only be generated either at the beginning of the period or the end of the period depending on the configuration, eliminating the randomness of earlier implementation.
- Removed redundant child table for saving invoices. Added a link field in Sales and Purchase Invoice

marination committed:
fix: Unlink and delete batch created from stock reco on cancel

(cherry picked from commit fc353231065f1c33890d5b7e55c2e4e2506d986d)

marination committed:
chore: Tests for Stock Entry

(cherry picked from commit a144548db95ec829b953ca6305fe898b1a9cdeb0)

marination committed:
chore: Tests for Stock Reconciliation

(cherry picked from commit 5bc5af1066e0bb4bead98598ed4bf9911d340734)

marination committed:
fix: Item Alternative Test

(cherry picked from commit 964de1fc69b32d377f8c2a6f15d6ab7381669606)

hrwX committed:
refactor: subscriptions

hrwX committed:
refactor: subscriptions

hrwX committed:
Merge branch 'develop' into subscription_refactor

barredterra commented on erpnext/accounts/doctype/subscription/subscription.json:
```diff
@@ -263,11 +250,23 @@
+   "link_doctype": "Sales Invoice",
+   "link_fieldname": "subscription"
+  }
+ ],
```
```suggestion
 "links": [
  {
   "group": "Buying",
   "link_doctype": "Purchase Invoice",
   "link_fieldname": "subscription"
  },
  {
   "group": "Selling",
   "link_doctype": "Sales Invoice",
   "link_fieldname": "subscription"
  }
 ],
```

hrwX committed:
fix: test cases for subscription

hrwX committed:
Merge branch 'subscription_refactor' of github.com:hrwx/erpnext into subscription_refactor

hrwX committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into subscription_refactor

hrwX committed:
fix: test cases and add type hints

hrwX committed:
Merge branch 'develop' into subscription_refactor

hrwX committed:
fix: grace period calculation

hrwX committed:
Merge branch 'develop' into subscription_refactor

barredterra commented on erpnext/accounts/doctype/subscription/subscription.py:
```diff
@@ -2,36 +2,48 @@
 		self.update_subscription_period(self.start_date)
 
-	def update_subscription_period(self, date=None, return_date=False):
+	def update_subscription_period(
```
Instead of the `return_date` flag, we could extract a method `get_subscription_period()` that can be called independently.

barredterra commented on erpnext/accounts/doctype/subscription/subscription.py:
```diff
@@ -101,286 +124,276 @@ def get_current_invoice_end(self, date=None):
 
-	def generate_invoice(self, prorate=0):
+	def generate_invoice(
+		self, from_date: Optional[bool] = None, to_date: Optional[bool] = None
```
```suggestion
		self, from_date: Optional[date] = None, to_date: Optional[date] = None
```

barredterra commented on erpnext/accounts/doctype/subscription/subscription.py:
```diff
@@ -101,286 +124,276 @@ def get_current_invoice_end(self, date=None):
 
-	def create_invoice(self, prorate):
+	def create_invoice(
+		self, from_date: Optional[bool] = None, to_date: Optional[bool] = None
```
```suggestion
		self, from_date: Optional[date] = None, to_date: Optional[date] = None
```

hrwX committed:
chore: apply code review changes

Co-authored-by: Raffael Meyer <14891507+barredterra@users.noreply.github.com>

govindsmenokee committed:
fix: HSN-wise-summary of outward supplies Updated Report

Report changes done in order to meet the specification as per govt guideline - [GUIDELINE](https://taxguru.in/goods-and-service-tax/12-points-note-filing-gstr-1-01st-2021-onwards.html)

(cherry picked from commit 363752510ead7d3b86693d3659b2157753f2762d)

# Conflicts:
#	erpnext/regional/report/hsn_wise_summary_of_outward_supplies/hsn_wise_summary_of_outward_supplies.py

hrwX committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into subscription_refactor

hrwX committed:
Merge branch 'subscription_refactor' of https://github.com/hrwX/erpnext into subscription_refactor

hrwX committed:
fix: review changes

sam-anam1 commented:
Hi all.

please in need help regarding this matter [post it ](https://discuss.erpnext.com/t/cancelling-subscriptions/87170)



Would like to use subscriptions with web shared clients
but the problem is when we cancelling subscriptions it will generate invoice
as the documentation

> If you are cancelling an active subscription, an invoice will immediately be generated. 
The generated invoice will be on pro-rata basis by default. If you want ERPNext always create an invoice for the full amount, uncheck the ‘Prorate’ field in Subscription Settings.
[Cancelling Subscriptions](https://docs.erpnext.com/docs/v12/user/manual/en/accounts/articles/how-to-manage-subscriptions-with-erpnext#cancelling-subscriptions)

it would be great has an option for that "no creating invoices when Cancelling & Restarting".
It dose make sense to me that not even make invoice when “uncheck Generate Invoice At Beginning Of Period” maybe.

However, the reason here as mention as example shared hosting plans
when customer cancel subscription due ICANN or even breaking rules such copyright or requesting for cancellation to changing domain extension etc… there is no way for refund or pro-rata basis… or full amount… so we wouldn’t wanna create an invoice or such.

Is there any side script or anyway to

Disable generating SI on Manually Cancelling and Restarting Subscription?
Disable Subscription Processing In The Background
or even other solution if it could change the manually status
We just wouldn’t go with custom code or doctype to avoid ERPNext updates

Many thanks!!

deepeshgarg007 commented:
@hrwX is this complete? Can you update the description and resolve the conflicts?

federicocalvo commented:
Hi, I don't know if it should be taken out as I don't see the detail of what the refactor looks like.

Something that conceptually should be possible to do, is to choose the payment gateway in the subscription, and not in the plan.

It is good that the option of the gateway is in the plan, but if I have 5 gateways, it forces me to make 5 plans for a product, since it is the user who chooses the payment method.

I just wanted to know if you raised it or should I create an issue outside.

hrwX committed:
fix: patch for migrating invoices

hrwX commented:
@sam-anam1 @federicocalvo This PR is a refactor for the existing functionality, I won't be adding any new functionality in this PR.
It would be great if you create a Github issue and tag it as improvement.

hrwX commented:
> improvement

I am in the process of doing it, will ping you once the PR is ready

hrwX committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into subscription_refactor

federicocalvo commented:
@hrwX done! #30571

nameduser0 commented:
@hrwX does this PR fix the issue someone told me about (I haven't checked myself) that if you create a subscription on 2022-03-15 the next invoice will be generated on 2022-05-01 and misses out 2022-04-01?

Is `Generate Invoice At Beginning Of Period` the difference between billing in advance and billing in arrears?
It's also not clear if this is the same thing as `Follow Calendar Months`?

Also the documentation mentions `Generate Invoice Days Early` which doesn't appear to exist.

Also, is this going into v13 too?

sam-anam1 commented:
> @sam-anam1 @federicocalvo This PR is a refactor for the existing functionality, I won't be adding any new functionality in this PR.
> It would be great if you create a Github issue and tag it as improvement.

@hrwX here it is [30640](https://github.com/frappe/erpnext/issues/30640)

deepeshgarg007 commented:
@hrwX is this ready for review? Please reopen once you start working on this again

deepeshgarg007 committed:
fix: Allow multi-currency opening invoices

ruthra-kumar committed:
feat: add payment terms status rpr to Selling homepage

deepeshgarg007 committed:
fix: Consistent accounting dimensions across Sales and Purchase docs

deepeshgarg007 committed:
fix: Ignore custom field validation while setup

deepeshgarg007 committed:
chore: Remove domains from Setup

deepeshgarg007 committed:
refactor: Remove education domain from ERPNext

deepeshgarg007 committed:
chore: remove files related to LMS

deepeshgarg007 committed:
chore: clean up hooks

deepeshgarg007 committed:
chore: Remove course links from training doctypes

deepeshgarg007 committed:
chore: Remove references for Assessment Group

deepeshgarg007 committed:
chore: Remove Fees Doctype

deepeshgarg007 committed:
chore: Remove references for Student doctype

deepeshgarg007 committed:
chore: Add patch to delete education doctypes

hrwX committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into subscription_refactor

hrwX committed:
chore: add patch to patches.txt

hrwX committed:
chore: add patch to patches.txt

deepeshgarg007 committed:
chore: Clean up patches

deepeshgarg007 committed:
fix: Multi currency opening invoices

ruchamahabal committed:
fix(UX): record reason for skipping attendance or marking absent for auto attendance (#30844)

deepeshgarg007 committed:
fix: Cost center filter on payment reconciliation

deepeshgarg007 committed:
fix: Vat Audit report fixes

(cherry picked from commit d5319a48269a4277818486e7ed74468e9107b8f4)

Altizo committed:
fix: update translate

rmehta committed:
fix(minor): discount accounting patch

rmehta committed:
fix(minor): fix create_accounting_dimensions_in_orders.py

sersaber committed:
feat: Copy task color from project template (#30857)

Co-authored-by: Rucha Mahabal <ruchamahabal2@gmail.com>

ankush committed:
chore: correct link

ankush committed:
fix: convert default_item_manufacturer to link field (#30835)

rmehta committed:
fix(minor): update frappe.error_log to new API (#30864)

* fix(minor): update frappe.error_log to new API

* refactor: changes for updated log_error api

Co-authored-by: Ankush Menat <ankush@frappe.io>

ankush committed:
fix(UX): misleading stock entry lables (#30870)

* fix(UX): misleading stock entry lables

* chore: field labels

[skip ci]

Co-authored-by: Marica <maricadsouza221197@gmail.com>

Co-authored-by: Marica <maricadsouza221197@gmail.com>

deepeshgarg007 committed:
test: Add test for payment reconciliation

deepeshgarg007 committed:
fix: Supply type for overseas invoices with payment of tax

nextchamp-saqib committed:
fix: supply type for sez invoices with payment of tax

maharshivpatel committed:
fix(india): e-invoice generation for registered composition gst category type (#30814)

fproldan committed:
fix: supplier quotation comparission report button

ruchamahabal committed:
fix: payment days calculation for employees joining/leaving mid-month (#30863)

deepeshgarg007 committed:
fix: Ignore loan repayments made from salary slip

(cherry picked from commit b7e1d40e43c136f0e967111e10af95ddc80e2fd5)

marination committed:
fix: Wrap SLE actual_qty in `flt` to avoid NoneType operation

- Since Batch cancellation SLEs do not set qtys (will fix separately), `merge_similar_entries` gets `actual_qty` as None
- This causes NoneType operation error on tests that cancel batch-serial reco
- Modified tests to avoid using commit and rollback explicitly

(cherry picked from commit d53228b153437f9dedcb1229bf579411f3122729)

marination committed:
fix: Remove commit from stock entry test. The assertion is not important

(cherry picked from commit c449b35cc11615fc7d9e616c2038b9e90c3957d6)

ankush committed:
fix: show group warehouse in Sales Order (#30891)

deepeshgarg007 committed:
fix: Consider paryt and party type as well in group by consolidated view

deepeshgarg007 committed:
fix: Show linked time sheets in sales invoice dashboard

barredterra committed:
feat: remove website from bank account

barredterra committed:
feat(German Translation): translate "Party" as "Partei"

adityahase committed:
chore: Remove reference to Instructor doctype (#30903)

References:
https://github.com/frappe/erpnext/pull/30830
https://github.com/frappe/erpnext/commit/44c54f69f83e28fe205eb5f0e5dc4f161a8c0bba

ankush committed:
test: activation

marination committed:
style: Spaces to Tabs

(cherry picked from commit a2fff8741e7ea928c4152d8f867db3ef9eb08b19)

ankush committed:
fix: disable form save on naming series tool (#30909)

maharshivpatel committed:
fix(india): keyerror while generating e-way bill from an e-invoice (#30879)

ankush committed:
fix: double future qty updates

update_qty_in_future_sle is reprocessing rows which are already
processed by process_sle_against_current_voucher

ankush committed:
chore: remove datettime formatting from debug report

This hides some information that would otherwise help during debugging

ruchamahabal committed:
feat(minor): fetch default salary structure and base from Employee Grade in Salary Structure Assignment (#30918)

maharshivpatel committed:
feat(india): store e-way bill auto calculated distance in sales invoice (#30908)

maharshivpatel committed:
feat(india): cancel e-way bill is enabled with e-invoicing APIs. (#30888)

deepeshgarg007 committed:
chore: Remove extra columns

(cherry picked from commit b0e929b8ae8780ada76ea6b699a2b4d840b31487)

# Conflicts:
#	erpnext/regional/report/hsn_wise_summary_of_outward_supplies/hsn_wise_summary_of_outward_supplies.py

deepeshgarg007 committed:
chore: Remove print statement

(cherry picked from commit a8fbd2451b498ad49356cf0e6a710616ab802465)

ankush committed:
chore: correct version in hooks

[skip ci]

deepeshgarg007 committed:
chore: Resolve conflicts

ankush committed:
fix: sort before picking next stock reco

ruchamahabal committed:
fix: remove check for already allocated earned leaves (#30931)

* fix: remove check for already allocated earned leaves

* fix: do not set New Leaves Allocated field as read-only for earned leaves

- removing this until there's a better way to update existing allocations

rohitwaghchaure committed:
fix: allow to use formatting for the field to_discuss in opportunity

nextchamp-saqib committed:
fix(pos): creating pos returns resets pricing rules & discounts (#30935)

HarryPaulo committed:
fix: subtract change_amount from paid_amount field on POS Register (#30922)

maharshivpatel committed:
fix(india): re-arrange e-way bill dialog fields (#30920)

ankush committed:
refactor: tax rule validity query (#30934)

ruchamahabal committed:
feat(HR): Leave Type configuration to allow over allocation (#30940)

nextchamp-saqib committed:
feat(india): generate qrcode button for e-invoice (#30939)

nextchamp-saqib committed:
fix(india): invoice type for a debit note e-invoice (#30948)

gavindsouza committed:
fix(charts): Pass fieldtype for chart data in selling reports

barredterra committed:
Merge branch 'develop' into subscription_refactor

hrwX commented:
I effed up this branch, reopening a new one here #30963
